### ISSUE TITLE
Inconsistent value of Pi in the source

### ISSUE THREAD

**Issue by serguei-patchkovskii** [Reference-LAPACK/lapack#428]
Hi,

I am trying to track down a numerical inconsistency in a real*16 (quadruple-precision) build of LAPACK, and I noticed that 2*Pi and Pi/2 is hard-coded in multiple locations throughout the source. The two values I see are:

PIOVER2, which is set to 1.57079632679489662 (either D0 or E0, depending on the precision) and

TWOPI, which is set to 6.2831853071795864769252867663 (again either D0 or E0). Parenthetically, this value is off in the last decimal digit - it should have been ...7666 if properly rounded.

These values are sufficiently accurate and consistent for real*4 and real*8 math. The value of PIOVER2 is already not quite accurate enough for real*10 math, and neither value is sufficiently accurate for real*16, which requires about 34 decimal digits.

The fix is to set

PIOVER2 = 1.570796326794896619231321691639751442099

TWOPI = 6.283185307179586476925286766559005768394

These values are given to 40 significant digits, which should be more than adequate for real*16 and complex*32 use.

The PIOVER2 value is found in:

./SRC/cbbcsd.f:     $                     ONE = 1.0E0, PIOVER2 = 1.57079632679489662E0,
./SRC/dbbcsd.f:     $                     ONE = 1.0D0, PIOVER2 = 1.57079632679489662D0,
./SRC/sbbcsd.f:     $                     ONE = 1.0E0, PIOVER2 = 1.57079632679489662E0,
./SRC/zbbcsd.f:     $                     ONE = 1.0D0, PIOVER2 = 1.57079632679489662D0,
./TESTING/EIG/cckcsd.f:     $                     PIOVER2 = 1.57079632679489662E0,
./TESTING/EIG/ccsdts.f:      PARAMETER          ( PIOVER2 = 1.57079632679489662E0,
./TESTING/EIG/dckcsd.f:     $                     PIOVER2 = 1.57079632679489662D0,
./TESTING/EIG/dcsdts.f:      PARAMETER          ( PIOVER2 = 1.57079632679489662D0,
./TESTING/EIG/sckcsd.f:     $                     PIOVER2 = 1.57079632679489662E0,
./TESTING/EIG/scsdts.f:      PARAMETER          ( PIOVER2 = 1.57079632679489662E0,
./TESTING/EIG/zckcsd.f:     $                     PIOVER2 = 1.57079632679489662D0,
./TESTING/EIG/zcsdts.f:      PARAMETER          ( PIOVER2 = 1.57079632679489662D0,

The TWOPI value is found in:

./SRC/clarnv.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./SRC/dlarnv.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./SRC/slarnv.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./SRC/zlarnv.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./TESTING/MATGEN/clarnd.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./TESTING/MATGEN/clatms.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./TESTING/MATGEN/clatmt.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./TESTING/MATGEN/dlarnd.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./TESTING/MATGEN/dlatms.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./TESTING/MATGEN/dlatmt.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./TESTING/MATGEN/slarnd.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./TESTING/MATGEN/slatms.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./TESTING/MATGEN/slatmt.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663E+0 )
./TESTING/MATGEN/zlarnd.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./TESTING/MATGEN/zlatms.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )
./TESTING/MATGEN/zlatmt.f:      PARAMETER          ( TWOPI = 6.2831853071795864769252867663D+0 )

**Comment by moe123** [comment]
Hello. you have C double exported constants: https://github.com/moe123/macadam/blob/master/macadam/mcconsts.h and in and in your system <math.h> you may relay on them. You can  also export real16 or new ones using https://www.wolframalpha.com. However .7666 vs .7663 won't change anything it will be rounded as normalized by rt.

**Comment by serguei-patchkovskii** [comment]
> Hello. you have C double exported constants here: https://github.com/moe123/macadam/blob/master/macadam/mcconsts.h and in your system <math.h> you may relay on them. You can also export real16 or new ones using https://www.wolframalpha.com. However ...7666 vs ...7663 won't change anything it will be rounded as normalized by rt.

I agree that the values in the code are OK for 32- and 64-bit IEEE reals (even if aesthetically jarring). They are however a problem for a 128-bit (quadruple-precision) real type. Which happens to be the case I am interested in.

**Comment by moe123** [comment]
Hello, yes I understand, meanwhile, I am not confident with 128-bit defs and its support. After checking in and out those concerns, the most straightforward  step I see, would be to use a compiler implementing or emulating such binary storage. Then computing estimate of each values needed, dumping to their hexa representation. I am not aware of what fortran today support for const declaration. Anyhow it departs from f77 compatibility forcibly.

**Comment by serguei-patchkovskii** [comment]
> Hello, yes I understand, meanwhile, I am not confident with 128-bit defs and its support. After checking in and out those concerns, the most straightforward step I see, would be to use a compiler implementing or emulating such binary storage. Then computing estimate of each values needed and dumping them to their respective hexadecimal true representation. I am not aware of what fortran today support for const declaration. Anyhow it departs from f77 compatibility forcibly.

I really do not understand what the concern is here, or why do you think one would have to use a "hexadecimal true representation". It appears that two issues have been conflated:
1. The existence of 128-bit (quadruple precision) real support
2. The standard conformance of the code specifying high-precision constants and using high-precision reals.

Concerning #1, 128-bit reals have been supported by Fortran compilers for at least 25 years, and quite possibly longer (The first time I used 128-bit reals, albeit not IEEE 128-bit reals, already with xlf on IBM AIX in 1994). Any recent version of GNU Fortran or Intel Fortran support extended IEEE 128-bit real variables. The performance of this floating-point representation is not great, but they are fast enough to be useful for many things. Including things where LAPACK would be helpful.

Concerning #2, I am aware of literally -nothing- in the Fortran standard that prohibits high-precision reals. In fact, Fortran says very little about the actual representation and precision of the floating-point values. One can certainly have a standard-conforming Fortran-77 compiler, where the default real is 64-bit, and the default double precision is 128 bit. Or 128 bit and 256 bit, for that matter [Strictly speaking, having a 128-bit real AND double precision would violate the part of the Fortran-77 standard dealing with EQUIVALENCE and sequence association through COMMON - but as far as I know, nothing in LAPACK relies on these ancient quirks of the standard]. It is also absolutely OK under the standard to specify constants with more digits than can be represented in the underlying type; the compiler is expected to convert the value the best it could. There is absolutely no need whatsoever to use any non-standard features or representations here. This is the beauty of Fortran, which does not dictate a particular underlying implementation, apart from some rather mild constraints.

For example, using gfortran 7 (or more recent), command-like options "-real-4-real-8 -real-8-real-16" will give you exactly this situation: the default real type will be IEEE 64-bit, and double precision will become the 128-bit extended IEEE type. As far as I know, this setup conforms to the Fortran-77 rules. It is also very useful, and one would like to have the correctly-working LAPACK in this situation.

To summarize, having 128-bit real support is useful, and as far as I can tell fully conforms to the Fortran standard. Such support is also ubiquitous.

**Comment by moe123** [comment]
"hexadecimal true representation" will give you a true normalized representation. There is nothing conflated here. AIX is an OS what's important is 128-bit floating-point hardware support or emulation. You know I pretty regular guy; I like synthesized chitchat.

**Comment by zerothi** [comment]
I think there is no reason _not_ to have the precision increased in case users ask for it. @serguei-patchkovskii could you make a PR for this?

**Comment by langou** [comment]
Thanks Nick. Thanks Sergei. This is merged.
